Sort on multiple fields

dstopp opened the pull request:
- Fixes #362
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - Database used: opus3_test_200211
  - All Django tests pass: NA
- Were any JavaScript or CSS files modified? Y
  - JSHINT run on all affected files: Y
  - Tested on Chrome, Firefox, Safari, and iOS: Yes
    (Remember to test all browser sizes)
  - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
-Add ability to sort on multiple fields:
  -add a 'plus' indicator to allow user to add a selected metadata field to sort on;
  -click on table header sort arrow will ask the user if they want to modify or append to current sort order behaviour
  -Limit total number of sortable fields to 9; disable button when limit is reached
-Add ability to reorder the sort fields
  -only rerender if the sort order was actually modified
-Add a small sort order indicator adjacent to the sort arrow that indicates the order in sorting
-Keep the last word of the header + sort arrow + sort number on same line
-Responsive behaviour on small screens
-Apply to both browse and cart pages
-Allow metadata fields in the sort order category that are not currently selected metadata fields to view <cannot add them but they will not disappear from the sort order>

Known problems:
 -As there is ongoing responsive work for mobile devices in the pipeline, this pull request may need to be updated w/those sorts of changes
-on Safari, small screen, lots of metadata fields in 'add' menu, i noticed that the scroll bar doesn't appear unless you try to add.  Not sure how to fix that.

dstopp committed:
-make the pills sortable
-on sorting pills, re-render the browse tab

dstopp committed:
-remove requirement of only one sort column in addtion to opusid;
-make the pill and table header click action a single function;
-minor code standarization

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 362_sort_on_multiple_fields

dstopp committed:
moved methods for sorting the metadata into a separate file

dstopp committed:
-more rename of classes to fit code standard;
-add the 'plus' at end of the sort pills
-create the dropdown menu for the not yet sorted metadata colums to be selected
-on click of a metadata item to add to sort, close the dropdown and fetch the data w/the new sort
-rename hideMenu to hideMenus because both need to be hidden for same reasonse; sprinkle hideMenus onclick

dstopp committed:
-added tooltip to the table header arrow;
-added a little superscript number next to a column indicated sort order;
-on click of the sort icon in the table header, if there are are two or more <different from clicked> fields, a confirm modal will ask to add or overwrite existing sort criteria
-

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 362_sort_on_multiple_fields

dstopp committed:
-remove sort icon from right and add diff one to left

dstopp committed:
-update text in confirm modal
-disable sort addition when there are no more metadata fields to add.
-adjust font and spacing on the pills
-added message to the sort list icon regarding sorting

dstopp committed:
-adjustements for responsiveness
-close menu at appropriate times
-add order number of sort next to items in sort
-make sure the last word stays with the order indicator on a line
-limit the number of items to sort on to 9

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 362_sort_on_multiple_fields

dstopp committed:
minor jshint fixes

dstopp committed:
-fix for duplicate sort pill

dstopp committed:
-fix for sort stop

dstopp committed:
Add a 'hover on' message to the '+' when the max. number of 9 fields to sort on has been added.

dstopp committed:
typo

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -114,10 +114,10 @@
-                            </span>&nbsp;
-                            <ul class="sort-contents list-inline"></ul>
+                            </span>
+							<div class="op-sort-contents"></div>
```
Bad indent

rfrenchseti replied:
This is still the wrong indent (although a different one)

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -183,8 +183,8 @@ <h2>Your search produced no results</h2>
-                                </span>&nbsp;
-                                <ul class="sort-contents list-inline"></ul>
+                                </span>
+								<div class="op-sort-contents"></div>
```
Bad indent

rfrenchseti replied:
This is still the wrong indent (although a different one)

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -278,6 +278,12 @@ <h2 class="dropdown-header"></h2>
+        <!-- dropdown menu for sort metadata columns -->
+        <div id="op-add-sort-metadata" class="dropdown-menu dropdown-menu-sm" data-toggle="dropdown" aria-labelledby="op-add-sort-metadata" aria-haspopup="true" aria-expanded="false">
+            <h2 class="dropdown-header">Add metadata field to sort</h2>
+			<div class="op-sort-list"></div>
```
Bad indent

rfrenchseti replied:
Still bad indent

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -316,7 +322,7 @@ <h3 class='op-header-text'></h3>
 
         <!--modal dialog for the metadata selector -->
-        <div id="op-select-metadata" class="modal fade" role="dialog" tabbindex="-1" data-backdrop="false" aria-labelledby="op-select-metadataLabel" aria-hidden="true">
+        <div id="op-select-metadata" class="modal fade" role="dialog" tabbindex="-1" data-backdrop="false" aria-labelledby="op-select-metadataLabel" ariaaria-hidden="true">
```
"ariaaria"

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -1134,13 +1151,14 @@ a:-webkit-any-link, a::any-link {
-#cart .thumb-page {
-    padding: 0 1.0em;
+#cart .op-cart-gallery-side {
+    /* margin-top:.25em; */
```
Remove comment

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1308,6 +1156,10 @@ var o_browse = {
 
         let suppressScrollY = false;
 
+        if (opus.prefs.order.length === 0) {
```
This does nothing...

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1532,14 +1382,39 @@ var o_browse = {
+            let lastWord = headerArr.pop();
+            let spacing = headerArr.length ? "&nbsp;" : "";
+            let lastWordWrappingGroup = `${headerArr.join(" ")}${spacing}<span class="op-last-word-group">${lastWord}` +
+                            `<span data-sort='${columnSorting}' ${orderToolTip} class='op-column-ordering fas fa-sort${icon}'>${columnOrderPostion}</span>`;
```
Indent?

rfrenchseti replied:
Also we normally use double quotes in HTML, and you can do that here since you're using template strings.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1532,14 +1382,39 @@ var o_browse = {
+                            `<span data-sort='${columnSorting}' ${orderToolTip} class='op-column-ordering fas fa-sort${icon}'>${columnOrderPostion}</span>`;
+            let columnOrdering = `<a href='' data-slug='${slug}' data-label='${label}'>${lastWordWrappingGroup}</a>`;
 
             $(`${tab} .op-data-table-view thead tr`).append(`<th id='${slug} 'scope='col' class='sticky-header'><div>${columnOrdering}</div></th>`);
```
Use double quotes in HTML

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -859,6 +860,10 @@ var opus = {
                             break;
+                        case "op-overwrite-sort-order":
+                            let addToSort = false;
+                            o_sortMetadata.onClickSortOrder($(`#${target}`).data("slug"), addToSort);
```
Why have a variable addToSort?

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -891,6 +896,10 @@ var opus = {
                             break;
+                        case "op-overwrite-sort-order":
+                            let addToSort = true;
+                            o_sortMetadata.onClickSortOrder($(`#${target}`).data("slug"), addToSort);
```
Why have a variable addToSort?

rfrenchseti replied:
This one is still there

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+        // click table column header to reorder by that column
+        // On click, if there are already 2 items in the sort <opusId + slug>, display overwrite modal option
+        // On ctrl click will append the selected sort
+        $("#browse, #cart").on("click", ".op-data-table-view th a",  function(e) {
```
Two space before function

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+            o_browse.hideMenus();
+            let inOrderList = $(this).find(".op-column-ordering").data("sort") !== "none";
+            let slug = $(this).data("slug");
+            if (inOrderList || opus.prefs.order.length < 2 || (e.ctrlKey || e.metaKey)) {
```
I think this is the logic you need it's just <=2

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+        $(".op-sort-contents").on("click", ".op-sort-order-add-icon", function(e) {
+            // allow the hover to work but the click appear to be disabled
+            if ($(".op-sort-order-add-icon").hasClass("op-sort-add-disabled")) {
+
```
Extra blank line

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+        $("body").addClass("op-prevent-pointer-events");
+        o_browse.showPageLoaderSpinner();
+
+        addToSort = (addToSort === undefined ? true : addToSort);
```
With ES6 you can use a default argument.... function(orderBy, addToSort=true).

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+        });
+
+        listHtml += `<div class="op-sort-order-add-icon list-inline-item" title="Edit sort by adding metadata fields">`+
+					   `<i class="fas fa-plus"></i>`+
```
Bad indent?

rfrenchseti replied:
This is still weirdly indented

dstopp replied:
The indent here is intentional; I am mimicking the html indent in js.

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+            }
+        });
+
+        listHtml += `<div class="op-sort-order-add-icon list-inline-item" title="Edit sort by adding metadata fields">`+
```
How about just "Add metadata fields to sort order"?

rfrenchseti commented on opus/application/static_media/js/sortMetadata.js:
```diff
@@ -0,0 +1,280 @@
+        // limit the total number of sort columns to 9
+        if (Object.keys(tableColumnFields).length === 0 || opus.prefs.order.length === 9) {
+            $(".op-sort-order-add-icon").addClass("op-sort-add-disabled");
+            $(".op-sort-order-add-icon").attr("title", "The maximun of 9 metadata fields to sort on has already been selected");
```
Change to: "The maximum of nine metadata sort fields has been reached"

rfrenchseti reviewed (changes requested):
- Always required OPUS ID to be the last item in the list.
- Put some kind of vertical divider in front of OPUS ID to make it clear there is something different going on.
- Make OPUS ID non-draggable.
- When there are <= 2 sort fields and the user clicks on a table header, do an "overwrite" without asking first. If there are >2 sort fields, do the current behavior.
- When there are no items to put in the dropdown list, the + has the wrong tooltip.
- I think it would be nice to add "Drag to reorder" to the tooltips for each of the individual pills.
- Currently the tooltip for a ^/v in the table header says "Change to ascending(descending) sort". Since you can click anywhere on the header, I think that tooltip should show up no matter where you hover. Also, I think the tooltip should change based on what's actually going to happen. If the field is already in the sort order, then it should say what it currently says. Right now, if the field is not in the sort order, there is no tooltip. It should say "Sort on this field" if there are currently <=2 sort fields or "Sort on this field or append to current sort<br>Ctrl+click to append" if there are >2 sort fields.
- I don't see the "+" being marked as inactive (greyed out) when it isn't going to do anything.
- It is possible for the "+" to wrap around and be on a line by itself. I think it should always stay with the last pill.
- All of our other modal-looking elements have title case in their header. Should we do that for this dropdown too?

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 362_sort_on_multiple_fields

# Conflicts:
#	opus/application/static_media/js/browse.js

dstopp committed:
Addressing pull request issues;
-force OPUS ID to always be the last in the list;
-add a vertical divider before OPUS ID
-make OPUS ID non-draggable
-Default behavior when there are only two items to sort on is to replace, not append
-Allow ctrl and shift to append to the sort w/out confirm modal
-Update the hint when there are no items left to add to the sort
-Make the '+' more obviously 'disabled' in appearance when disabled
-Change the title case in the modal header

dstopp committed:
-normalizeURL changes for order=
-remove tabs & typos in base.html
-update text for tooltip
-address code comments from the pull request

rfrenchseti reviewed (changes requested):
- Still need normalizeurl tests for missing opusid or opusid in the middle of the sort order.
- Think about moving + to the left of opusid and greying out and putting distance before opusid to make it look different than the other pills
- Make sure + always stays with the pill immediately to its left when wrapping

dstopp committed:
-minor tab issues, pull comment address

dstopp committed:
-updated help/getting started text
-update the tooltips on the table headers and make them appear on the entire clickable bit
-put 'drag to reorder' tooltip comment on second line

dstopp committed:
-Click to reset sort order on icon, include confirm modal
-Add 'Sort by' after icon to describe action of icon
-Move the '+' to before OPUS ID
Note - click on OPUS ID   NOT fixed yet.

dstopp committed:
-update message in confirm reset modal
-update unit tests for order
-add 3 new unit test for order

dstopp committed:
-fix some spacing and changing the  order of the opusId column

dstopp committed:
-this keeps the plus on the right; still working on not letting it move at all

dstopp committed:
-tie plus & opus together so they do not get affected by dragging things on sort

dstopp committed:
-fix for "clicking on OPUS ID in the table header messes things up"

dstopp committed:
add'l tests for dec

dstopp committed:
-prevent ctrl-click when there are 9 or more sort items;
-disable '+' when there are 9 or more sort items <not just exactly 9>

dstopp committed:
update tooltip when there are 9 or more sort items

dstopp committed:
yet another tooltip mod

dstopp committed:
fix for updating the order when the object is not one of the selected metadata columns in the table
